src/mem3/README_reshard.md

```diff
@@ -0,0 +1,88 @@
+
+3. Job's internal splitting state. This state tracks the steps taken during shard splitting by each job. This state is mostly relevant for the `mem3_reshard_job` module. This state is kept in `#job{}`'s `split_state` field. The progression of these states is linear going from one state to the next. That's reflected in the code as well, when one state is done, `mem3_reshard_job:next_state/1` is called which returns the next state in the list. The list is defined in `mem3_reshard.hrl` file in `SPLIT_STATES` macro. This simplistic transition is also one of the reasons why a gen_fsm wasn't considered for `mem3_reshard_job` logic.
+
+Another interesting aspect is how the `split_state` transitions happen in `mem3_reshard_job`. What follows is an examination of that. A job starts running in the `new` state or from a previously checkpointed state. In the later case, the job goes through some recovery logic (see `?STATE_RESTART` macro in `mem3_reshard_job`) where it tries to resume its work from where it left of. It means, for example, if it was in the `initial_copy` state and was interrupted it might have to reset the target files and copy everything again. After recovery, the state execution logic is driven by `run(#job{})` which ends up calling `?MODULE:State(#job{})` state specific functions for each state.
+
+In `mem3_reshard_job:switch_state/2` job's history is updated, any current `state_info` is cleared, job state is switched in the `#job{}` record and then `mem3_reshard` asked to checkpoint this state. That notificaiton is done via gen_server cast message. After that is send the job process sits and waits. In the meantime `mem3_reshard` manager checkpoints, which mean it updates both its ets table with the new `#job{}` record and also persists the state with the `mem3_reshard_store`, and finally it notifies the job process that checkpointing is done by calling `mem3_reshard_job:checkpoint_done/1`.
+
+`mem3_reshard_job:checkpoint_done/1` function call then sends a `do_state` cast to the job's gen_server which proceeds to execute that particular state.
+
+Most states in `mem3_reshard_job` try not to block the gen_server process and instead launch worker processes to perform long running operations. It is usually just one worker process but it could be multiple as well. After that it mostly waits for the workers to finish and inspects their exit signal. If they all exit normal (see `mem_reshard:worker_exited/2`) it calls `switch_state/2` again with the next state and the whole process repeats until completion. Upon completion, the `mem3_reshard_job` gen_server exits normally.
```

Comment:
Is it possible for a very busy DB to simply never finish the resharding, falling behind on checkpoints due to resource starvation or simply the huge volume of db updates?

Reply:
Yeah that is a possibility, I'll mention it. It's a bit like compaction and in fact the main bulk copy of data is pretty much the same logic as the compactor just that it writes to multiple targets.